zio/csvio/reader.go

```diff
@@ -0,0 +1,109 @@
+package csvio
+
+import (
+	"encoding/csv"
+	"errors"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zson"
+)
+
+type Reader struct {
+	zctx      *zson.Context
+	reader    *csv.Reader
+	marshaler *zson.MarshalZNGContext
+	strings   bool
+	valid     bool
+	hdr       []string
+	vals      []interface{}
+}
+
+// XXX This is a placeholder for an option that will allow one to convert
+// all csv fields to strings and defer any type coercion presumably to
+// Z shapers.  Currently, this causes an import cycle because the csvio
+// Writer depends on fuse.  We should refactor this so whatever logic wants
+// to tack on a fuse operator happens outside of zio.  See issue #2315
+//type ReaderOpts struct {
+//	StringsOnly bool
+//}
+
+func NewReader(reader io.Reader, zctx *zson.Context) *Reader {
+	return &Reader{
+		zctx:      zctx,
+		reader:    csv.NewReader(reader),
```

Comment:
Avoid a slice allocation on every trip through `csv.Reader.Read`.
```suggestion
    csvReader := csv.NewReader(reader)
    csvReader.ReuseRecord = true
    return &Reader{
        zctx:      zctx,
        reader:    csvReader,
```

Reply:
nice